File: domain/Reaccionables/Expertise.py
from typing import List

from domain.Reaccionables.Puntuacion import Puntuacion


class Expertise:
    def __init__(self, texto: str, puntuaciones: List[Puntuacion]) -> None:
        self.puntuaciones: List[Puntuacion] = puntuaciones
        self.texto: str = texto

    def ser_puntuado(self, puntuacion: Puntuacion) -> None:
        ya_reaccionado = False
        for indexOfElem in range(len(self.puntuaciones)):
            if self.puntuaciones[indexOfElem].autor == puntuacion.autor:
                self.puntuaciones[indexOfElem] = puntuacion
                ya_reaccionado = True
        if not ya_reaccionado:
            self.puntuaciones.append(puntuacion)

    def ser_despuntuado(self, puntuacion: Puntuacion) -> None:
        if any(puntuacion.autor == elem.autor for elem in self.puntuaciones):
            self.puntuaciones.remove(puntuacion)

File: src/main/java/entities/usuarios/reaccionables/rasgos/Expertise.java
package entities.usuarios.reaccionables.rasgos;

import com.google.gson.annotations.Expose;
import db.EntidadPersistente;

import javax.persistence.*;
import java.util.List;

@Entity
@Table(name = "expertise")
public class Expertise extends EntidadPersistente {
    @ManyToOne @Expose
    private TipoExpertise tipoExpertise;
    @OneToMany @Expose
    private List<Puntuacion> puntuaciones;

    public Expertise(TipoExpertise tipoExpertise, List<Puntuacion> puntuaciones) {
        this.tipoExpertise = tipoExpertise;
        this.puntuaciones = puntuaciones;
    }

    public Expertise(){}

    public List<Puntuacion> getPuntuaciones() {
        return puntuaciones;
    }

    public void setPuntuaciones(List<Puntuacion> puntuaciones) {
        this.puntuaciones = puntuaciones;
    }

    public void serPuntuado(Puntuacion nuevaPuntuacion) {
        puntuaciones.removeIf(puntuacion ->
                puntuacion.getUsuarioPuntuador().equals(nuevaPuntuacion.getUsuarioPuntuador()));
        puntuaciones.add(nuevaPuntuacion);
    }

    public void serDespuntuado(Puntuacion puntuacion) {
        puntuaciones.remove(puntuacion);
    }

    public String getDescripcion() {
        return tipoExpertise.getDescripcion();
    }

    public TipoExpertise getTipoExpertise() {
        return tipoExpertise;
    }

    public void setTipoExpertise(TipoExpertise tipoExpertise) {
        this.tipoExpertise = tipoExpertise;
    }
}
